stat/distuv/poisson.go

```diff
@@ -0,0 +1,109 @@
+// Copyright ©2017 The gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package distuv
+
+import (
+	"math"
+	"math/rand"
+
+	"gonum.org/v1/gonum/mathext"
+)
+
+// Poisson implements the Poisson distribution, a discrete probability distribution
+// that expresses the probability of a given number of events occurring in a fixed
+// interval of time and/or space.
+// The poisson distribution has density function:
+//  f(k) = λ^k / k! e^(-λ)
+// For more information, see https://en.wikipedia.org/wiki/Poisson_distribution.
+type Poisson struct {
+	// Lambda is the average number of events in an interval.
+	// Lambda must be greater than 0.
+	Lambda float64
+
+	Source *rand.Rand
+}
+
+// CDF computes the value of the cumulative distribution function at x.
+func (p Poisson) CDF(x float64) float64 {
+	if x < 0 {
+		return 0
+	}
+	return mathext.GammaInc(math.Floor(x+1), p.Lambda) / math.Gamma(x+1)
+}
+
+// ExKurtosis returns the excess kurtosis of the distribution.
+func (p Poisson) ExKurtosis() float64 {
+	return 1 / p.Lambda
+}
+
+// LogProb computes the natural logarithm of the value of the probability
+// density function at x.
+func (p Poisson) LogProb(x float64) float64 {
+	if x < 0 {
+		return math.Inf(-1)
+	}
+	lg, _ := math.Lgamma(math.Floor(x) + 1)
+	return x*math.Log(p.Lambda) - p.Lambda - lg
+}
+
+// Mean returns the mean of the probability distribution.
+func (p Poisson) Mean() float64 {
+	return p.Lambda
+}
+
+// NumParameters returns the number of parameters in the distribution.
+func (Poisson) NumParameters() int {
+	return 1
+}
+
+// Prob computes the value of the probability density function at x.
+func (p Poisson) Prob(x float64) float64 {
+	return math.Exp(p.LogProb(x))
+}
+
+// Rand returns a random sample drawn from the distribution.
+func (p Poisson) Rand() float64 {
```

Comment:
I'm pretty sure this is not how we want to generate random variables. It is very expensive, with an expected  `Lambda` random numbers generated per random variable. For small lambda, I think we instead want to use the inverse transform sampling in the wiki article (better source: 4.1 of http://statweb.stanford.edu/~owen/mc/Ch-nonunifrng.pdf) . For large lambda, we definitely want to use rejection sampling with a Normal distribution. I don't know where that divide is, but I would guess it's at `Lambda > 10` or so. See page 511 of http://luc.devroye.org/chapter_ten.pdf, and note the discussion on the best choice of Delta

Reply:
I believe for the Numpy implementation (https://github.com/numpy/numpy/blob/master/numpy/random/mtrand/distributions.c) the divide is `Lambda >= 10` as seen in the `rk_poisson` function.